From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HttpRafa <rafael@kienitz.link>
Date: Sat, 10 Jun 2023 22:20:10 +0200
Subject: [PATCH] Added network core


diff --git a/build.gradle.kts b/build.gradle.kts
index 149f9088fe806467656e8b1c4157df60fda69ba7..c2b65c396293f84ba46645c857e4e7a7cf8e5e77 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -27,6 +27,20 @@ val annotationsVersion = "24.0.1"
 
 dependencies {
     // api dependencies are listed transitively to API consumers
+    
+    // Raa start
+    api("de.network.systems:core:" + findProperty("networkVersion")) {
+        isChanging = true
+    }
+    api("de.network.systems:server:" + findProperty("networkVersion") + ":remapped-" + findProperty("targetMapping")) {
+        isChanging = true
+    }
+
+    api("org.mongodb:mongodb-driver-sync:4.8.2")
+    api("io.lettuce:lettuce-core:6.2.3.RELEASE")
+    api("commons-io:commons-io:2.12.0")
+    // Raa end
+
     api("com.google.guava:guava:31.1-jre")
     api("com.google.code.gson:gson:2.10")
     api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.9") // Paper
diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index 72175dcae49f75b494ab70958053ed994a8828f4..4e4d3d5e7f2fb3751e9197054bfd23e15239f182 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -15,7 +15,7 @@ import org.jetbrains.annotations.Nullable;
  * player that is stored on the disk and can, thus, be retrieved without the
  * player needing to be online.
  */
-public interface OfflinePlayer extends ServerOperator, AnimalTamer, ConfigurationSerializable {
+public interface OfflinePlayer extends ServerOperator, AnimalTamer, ConfigurationSerializable, de.network.systems.core.player.OfflineNetworkPlayer { // Raa
 
     /**
      * Checks if this player is currently online
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 0b4f9dee48f8d49977dc9a5f104ee062ca2cbfae..df52e8a0ab2887b9504d3514ffcaff7b26aaf752 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -43,7 +43,7 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Represents a player, connected or not
  */
-public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient, de.network.systems.core.player.NetworkPlayer { // Raa
 
     // Paper start
     @Override
