From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HttpRafa <rafael@kienitz.link>
Date: Sat, 10 Jun 2023 23:17:13 +0200
Subject: [PATCH] Added option to fake secure chat


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 4ac3fa45cd155ae8a852e26d4d4d1f16b28efdc2..7c0d345ec18d177b594ed1a76b7b767e3003bfb6 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -261,6 +261,7 @@ public class GlobalConfiguration extends ConfigurationPart {
             }
         }
         public int maxJoinsPerTick = 5;
+        public boolean fakeSecureChat = false; // Raa
         public boolean fixEntityPositionDesync = true;
         public boolean loadPermissionsYmlBeforePlugins = true;
         @Constraints.Min(4)
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 9d46536f80b5b3e6641fd377c02166a431edfd77..7833c7e77b6b1abecd8ddb81337a970f1e3fea27 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2076,7 +2076,7 @@ public class ServerPlayer extends Player {
     }
 
     public void sendServerStatus(ServerStatus metadata) {
-        this.connection.send(new ClientboundServerDataPacket(metadata.description(), metadata.favicon().map(ServerStatus.Favicon::iconBytes), metadata.enforcesSecureChat()));
+        this.connection.send(new ClientboundServerDataPacket(metadata.description(), metadata.favicon().map(ServerStatus.Favicon::iconBytes), io.papermc.paper.configuration.GlobalConfiguration.get().misc.fakeSecureChat || metadata.enforcesSecureChat())); // Raa
     }
 
     @Override
